packages/backend/src/apps/m365-excel/common/oauth/token-request.ts

```diff
@@ -0,0 +1,91 @@
+import { randomUUID } from 'crypto'
+import FormData from 'form-data'
+import jwt from 'jsonwebtoken'
+
+import {
+  getM365TenantInfo,
+  type M365TenantInfo,
+} from '@/config/app-env-vars/m365'
+import type { IHttpClient } from '@/helpers/http-client'
+
+export interface MsGraphAccessToken {
+  value: string
+  requestTimestamp: number
+  expiryTimestamp: number
+}
+
+// https://learn.microsoft.com/en-us/entra/identity-platform/certificate-credentials
+function makeClientAssertion(tenant: M365TenantInfo): string {
+  const currTimeSeconds = Math.round(Date.now() / 1000)
+
+  const header = {
+    alg: 'RS256',
+    typ: 'JWT',
+    x5t: Buffer.from(tenant.clientThumbprint, 'hex').toString('base64url'),
+  }
+  const payload = {
+    jti: randomUUID(),
+
+    aud: `https://login.microsoftonline.com/${tenant.id}/oauth2/v2.0/token`,
+    iss: tenant.clientId,
+    sub: tenant.clientId,
+
+    iat: currTimeSeconds,
+    nbf: currTimeSeconds,
+    exp: currTimeSeconds + 600, // 10 minutes
```

Comment:
wah this one nice!! can i just check 10 minutes is the guideline m365 give right?

Reply:
yep thats right